// @wix/yoshi-flow-editor-runtime/build/esm/publicUtils.js
var warnInDevelopment = function(msg) {
    if (process.env.NODE_ENV === 'development') {
        console.warn(msg);
    }
};
var viewerModeEditorParams = {
    editorSDKSrc: null,
    componentId: null,
    instance: null,
};
export var getQueryParams = function(search) {
    if (typeof search === 'undefined' && typeof window !== 'undefined') {
        search = window.location.search;
    }
    return new URLSearchParams(search);
};
export var getPanelUrl = function(componentName, panelName, options) {
    var _a;
    if (options === void 0) {
        options = {};
    }
    var isLocal = process.env.NODE_ENV === 'development';
    if (isLocal) {
        return "https://editor.wixapps.net/render/local/panels/" + componentName + "/" + panelName + "?cdnPort=" + ((_a = options === null || options === void 0 ? void 0 : options.cdnPort) !== null && _a !== void 0 ? _a : 3200);
    }
    return "https://editor.wixapps.net/render/prod/panels/" + process.env.ARTIFACT_ID + "/" + process.env.ARTIFACT_VERSION + "/" + componentName + "/" + panelName;
};
export var getModalUrl = function(componentName, panelName, instance) {
    var _a;
    var isLocal = process.env.NODE_ENV === 'development';
    var urlPostfix = componentName + "/" + panelName;
    if (isLocal) {
        return "https://editor.wixapps.net/render/local/modals/" + urlPostfix + "?cdnPort=" + ((_a = process.env.CDN_PORT) !== null && _a !== void 0 ? _a : 3200) + "&instance=" + instance;
    }
    return "https://editor.wixapps.net/render/prod/modals/" + process.env.ARTIFACT_ID + "/" + process.env.ARTIFACT_VERSION + "/" + urlPostfix + "?instance=" + instance;
};
export var getEditorParams = function() {
    // We don't use it in viewer environment anywhere now, but since it's a public function,
    // we should prevent runtime errors and follow SDK render prop logic, where values is `null` in viewer mode.
    if (!isEditor()) {
        warnInDevelopment("You are using getEditorParams in viewer environment. Please, remove it from your Widget, since it is useful only for platform's Settings Panel component");
        return viewerModeEditorParams;
    }
    var queryParams = getQueryParams();
    return {
        editorSDKSrc: queryParams.get('sdkVersion') || queryParams.get('wix-sdk-version'),
        componentId: queryParams.get('componentId'),
        instance: queryParams.get('instance'),
    };
};
export var isEditor = function() {
    // We are checking SSR and CSR cases
    return typeof window !== 'undefined' && !!window.__EDITOR_MODE__;
};
